test/src/dgkernels/DGCoupledDiffusion.C

```diff
@@ -0,0 +1,133 @@
+/*           (c) 2010 Battelle Energy Alliance, LLC             */
+/*                   ALL RIGHTS RESERVED                        */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
+#include "DGCoupledDiffusion.h"
+
+// MOOSE includes
+#include "MooseVariable.h"
+#include "SystemBase.h"
+
+// libMesh includes
+#include "libmesh/utility.h"
+
+template <>
+InputParameters
+validParams<DGCoupledDiffusion>()
+{
+  InputParameters params = validParams<DGKernel>();
+  // See header file for sigma and epsilon
+  params.addRequiredParam<Real>("sigma", "sigma");
+  params.addRequiredParam<Real>("epsilon", "epsilon");
+  params.addParam<MaterialPropertyName>(
+      "diff", 1., "The diffusion (or thermal conductivity or viscosity) coefficient.");
+  params.addRequiredCoupledVar("v", "The governing variable that controls diffusion of u.");
+  return params;
+}
+
+DGCoupledDiffusion::DGCoupledDiffusion(const InputParameters & parameters)
+  : DGKernel(parameters),
+    _epsilon(getParam<Real>("epsilon")),
+    _sigma(getParam<Real>("sigma")),
+    _diff(getMaterialProperty<Real>("diff")),
+    _diff_neighbor(getNeighborMaterialProperty<Real>("diff")),
+    _v_var(*getVar("v", 0)),
+    _v(_v_var.sln()),
+    _v_neighbor(_v_var.slnNeighbor()),
+    _grad_v(_v_var.gradSln()),
+    _grad_v_neighbor(_v_var.gradSlnNeighbor()),
+    _v_id(coupled("v"))
+{
+}
+
+Real
+DGCoupledDiffusion::computeQpResidual(Moose::DGResidualType type)
+{
+  Real r = 0;
+
+  const unsigned int elem_b_order = _v_var.order();
+  const double h_elem =
+      _current_elem->volume() / _current_side_elem->volume() * 1. / Utility::pow<2>(elem_b_order);
+
+  switch (type)
+  {
+    case Moose::Element:
+      r -= 0.5 * (_diff[_qp] * _grad_v[_qp] * _normals[_qp] +
+                  _diff_neighbor[_qp] * _grad_v_neighbor[_qp] * _normals[_qp]) *
+           _test[_i][_qp];
+      r += _epsilon * 0.5 * (_v[_qp] - _v_neighbor[_qp]) * _diff[_qp] * _grad_test[_i][_qp] *
+           _normals[_qp];
+      r += _sigma / h_elem * (_v[_qp] - _v_neighbor[_qp]) * _test[_i][_qp];
+      break;
+
+    case Moose::Neighbor:
+      r += 0.5 * (_diff[_qp] * _grad_v[_qp] * _normals[_qp] +
+                  _diff_neighbor[_qp] * _grad_v_neighbor[_qp] * _normals[_qp]) *
+           _test_neighbor[_i][_qp];
+      r += _epsilon * 0.5 * (_v[_qp] - _v_neighbor[_qp]) * _diff_neighbor[_qp] *
+           _grad_test_neighbor[_i][_qp] * _normals[_qp];
+      r -= _sigma / h_elem * (_v[_qp] - _v_neighbor[_qp]) * _test_neighbor[_i][_qp];
```

Comment:
If you want to test just the off-diag machinery, you could do a much simpler residual. For example, you could drop the material property, sigma and epsilon. Then, this would be a simple test object.

However, this almost seems like it would be useful in general. If so, we can put this in moose/framework. But then, I'd say, we would need to improve doco for this class. Like add the doco for the class, its member variables and input parameters.

Reply:
Can you think of a PDE where one variable's governing equation contains the diffusion term of another variable? I can't think of one off the top of my head, so I'd be more inclined to go the simplification route that you suggest.

Reply:
Yeah, me neither, so let's go with just simplifying this test object.